src/driver.rs

```diff
@@ -60,10 +61,58 @@ fn test_arg_value() {
 }
 
 #[allow(clippy::too_many_lines)]
+
+struct ClippyCallbacks;
+
+impl rustc_driver::Callbacks for ClippyCallbacks {
+    fn after_parsing(&mut self, compiler: &interface::Compiler) -> bool {
+        let sess = compiler.session();
+        let mut registry = rustc_plugin::registry::Registry::new(
+            sess,
+            compiler.parse().expect(
+                "at this compilation stage \
+                    the crate must be parsed",
+            ).peek().span,
+        );
+        registry.args_hidden = Some(Vec::new());
+
+        let conf = clippy_lints::read_conf(&registry);
+        clippy_lints::register_plugins(&mut registry, &conf);
+
+        let rustc_plugin::registry::Registry {
+            early_lint_passes,
+            late_lint_passes,
+            lint_groups,
+            llvm_passes,
+            attributes,
+            ..
+        } = registry;
+        let mut ls = sess.lint_store.borrow_mut();
+        for pass in early_lint_passes {
+            ls.register_early_pass(Some(sess), true, false, pass);
+        }
+        for pass in late_lint_passes {
+            ls.register_late_pass(Some(sess), true, pass);
+        }
+
+        for (name, (to, deprecated_name)) in lint_groups {
+            ls.register_group(Some(sess), true, name, deprecated_name, to);
+        }
+        clippy_lints::register_pre_expansion_lints(sess, &mut ls, &conf);
+        clippy_lints::register_renamed(&mut ls);
+
+        sess.plugin_llvm_passes.borrow_mut().extend(llvm_passes);
+        sess.plugin_attributes.borrow_mut().extend(attributes);
+
+        // Continue execution
+        true
```

Comment:
The old code did `controller.compilation_done.stop = Compilation::Stop;`, isn't this a change in behavior? (Same thing as in Miri, which is the only reason I noticed this.)

Reply:
Stopping after compilation is done doesn't really do anything.

Reply:
I was mostly worried about Clippy accidentally emitting a binary (which a bug in Miri made it do for some time, and I actually have no idea what made it stop doing that). Not sure if that's related to stopping here.

Reply:
The previous (and new) code should emit a binary here.

Reply:
> Stopping after compilation is done doesn't really do anything.

This might be what's causing `cargo clippy` not do anything on a second run. We should investigate in the future, but this PR does the right thing in keeping the previous behaviour